src/DeviceInput/InputDevices/deviceSourceManager.ts

```diff
@@ -0,0 +1,308 @@
+        this._devices = new Array<Array<DeviceSource<DeviceType>>>(numberOfDeviceTypes);
+        this._firstDevice = new Array<number>(numberOfDeviceTypes);
+        this._deviceInputSystem = new DeviceInputSystem(engine);
+
+        this._deviceInputSystem.onDeviceConnected = (deviceName) => {
+            const deviceType = this._getDeviceTypeFromName(deviceName);
+            const deviceSlot = this._getDeviceSlot(deviceName);
+
+            this.onBeforeDeviceConnectedObservable.notifyObservers({ deviceType, deviceSlot });
+            this._addDevice(deviceName);
+            this.onAfterDeviceConnectedObservable.notifyObservers({ deviceType, deviceSlot });
+        };
+        this._deviceInputSystem.onDeviceDisconnected = (deviceName) => {
+            const deviceType = this._getDeviceTypeFromName(deviceName);
+            const deviceSlot = this._getDeviceSlot(deviceName);
+
+            this.onBeforeDeviceDisconnectedObservable.notifyObservers({ deviceType, deviceSlot });
+            this._removeDevice(deviceName);
+            this.onAfterDeviceDisconnectedObservable.notifyObservers({ deviceType, deviceSlot });
+        };
+
+        if (!this._deviceInputSystem.observeInput && enableObserveEvents) {
+            this._deviceInputSystem.observeInput = (deviceName, inputIndex, previousState, currentState) => {
+                const deviceType = this._getDeviceTypeFromName(deviceName);
+                const deviceSlot = this._getDeviceSlot(deviceName);
+
+                if (deviceType == DeviceType.Keyboard || deviceType == DeviceType.Mouse || deviceType == DeviceType.Touch) {
+                    this.getDeviceSource(deviceType, deviceSlot)?.onInputChangedObservable.notifyObservers({ inputIndex, previousState, currentState });
+                }
+            };
+        }
+    }
+
+    // Public Functions
+    /**
+     * Checks for current device input value, given DeviceType, slot, and inputIndex
+     * @param type Enum specifiying device type
+     * @param inputIndex Index of device input
+     * @param deviceSlot "Slot" or index that device is referenced in
+     * @returns Current value of input
+     */
+    public getInput<T extends DeviceType>(type: T, inputIndex: DeviceInputs<T>, deviceSlot: number = this._firstDevice[type]): Nullable<number> {
+        if (!this._devices[type] || this._firstDevice[type] === undefined || this._devices[type][deviceSlot] === undefined) {
+            return null;
+        }
+
+        return this._devices[type][deviceSlot].getInput(inputIndex);
+    }
+
+    /**
+     * Gets a DeviceSource, given a type and slot
+     * @param deviceType Enum specifiying device type
+     * @param deviceSlot "Slot" or index that device is referenced in
+     * @returns DeviceSource object
+     */
+    public getDeviceSource(deviceType: DeviceType, deviceSlot: number = this._firstDevice[deviceType]): Nullable<DeviceSource<DeviceType>> {

```

Comment:
This will not actually get us the type safety we are aiming for. We need a generic type parameter for the conditional type to work:
```suggestion
    public getDeviceSource<T extends DeviceType>(deviceType: T, deviceSlot: number = this._firstDevice[deviceType]): Nullable<DeviceSource<T>> {
```

Reply:
Changed